File: packages/platform-desktop/src/adapters/crypto.ts
// CryptoAdapter backed by the Rust process, not WASM. Every call is a Tauri command over
// IPC; the VEK lives in the shell and never enters this webview, which is the whole reason
// the desktop app is a Rust binary. See docs/desktop-port.md.
//
// The core's structs already serialize camelCase, so the command results land in the shapes
// @core/adapters/crypto declares and need no mapping.

import type {
	CryptoAdapter,
	EncryptedPayload,
	KdbxRawEntry,
	OpenKdbxInput,
	PasskeyAssertion,
	PasskeyImportResult,
	PasskeyRegistration,
	PasswordSlotBlob,
	UnwrapPasswordSlotInput,
	UnwrapWebauthnSlotInput,
	VekEncrypted,
	VerifyPasswordSlotInput,
	VerifyWebauthnSlotInput,
	WrapPasswordSlotInput,
	WrapWebauthnSlotInput,
} from "@core/adapters/crypto";
import { invoke } from "@tauri-apps/api/core";
import { markLocked, markUnlocked, onExternalChange, onExternalLock } from "./vault-session";

/**
 * `magicVersion` is a Uint8Array in the adapter contract; Tauri's IPC is JSON, and the
 * command signature takes `Vec<u8>`, so it rides as a plain number array.
 */
function bytes(v: Uint8Array): number[] {
	return Array.from(v);
}

/** Phase 0 gaps. These throw rather than silently no-op so a wired-up UI fails loudly. */
function notWired(what: string): never {
	throw new Error(`${what} is not wired on desktop yet`);
}

export const desktopCrypto: CryptoAdapter = {
	// Single-process, single active vault for now, so there is no per-vault VEK map to
	// bind to (the extension needs one because its background serves every open view).
	// See docs/multiple-vaults.md "Per-vault VEK".

	// Every path that leaves the VEK loaded reports the transition, and only after the Rust
	// side confirms: roster sync runs on this signal, and a merge cannot decrypt anything
	// while locked. buildCryptoAdapter fires the same four for the platforms built on it;
	// this adapter is hand-written, so the four are marked individually. See ./vault-session.
	generateVek: async () => {
		const vek = await invoke<string>("crypto_generate_vek");
		markUnlocked();
		return vek;
	},
	unlockWithVek: async (vekB64) => {
		await invoke<void>("crypto_unlock_with_vek", { vekB64 });
		markUnlocked();
	},
	exportVek: () => invoke<string>("crypto_export_vek"),
	rotateVek: () => invoke<string>("crypto_rotate_vek"),
	lock: async () => {
		await invoke<void>("crypto_lock");
		markLocked();
	},
	isLocked: () => invoke<boolean>("crypto_is_locked"),

	// Sync merges write entries out of band, so the open list has to be told. A lock from
	// outside this webview (the spotlight window, an idle timeout in the shell) does not exist
	// yet; when it does it will emit a Tauri event that calls markLocked.
	onExternalLock,
	onExternalChange,

	generateSalt: () => invoke<string>("crypto_generate_salt"),
	generateSlotId: () => invoke<string>("crypto_generate_slot_id"),

	wrapVekPassword: (input: WrapPasswordSlotInput) =>
		invoke<PasswordSlotBlob>("crypto_wrap_vek_password", {
			password: input.password,
			saltB64: input.saltB64,
			slotIdB64: input.slotIdB64,
			magicVersion: bytes(input.magicVersion),
		}),
	// The ordinary unlock: a wrong password returns false rather than throwing, so the
	// transition is reported on the result, not on the call completing.
	unwrapVekPassword: async (input: UnwrapPasswordSlotInput) => {
		const ok = await invoke<boolean>("crypto_unwrap_vek_password", {
			password: input.password,
			saltB64: input.saltB64,
			slotIdB64: input.slotIdB64,
			verifierB64: input.verifierB64,
			wrapIvB64: input.wrapIvB64,
			wrappedVekB64: input.wrappedVekB64,
			magicVersion: bytes(input.magicVersion),
		});
		if (ok) markUnlocked();
		return ok;
	},
	verifyPasswordSlot: (input: VerifyPasswordSlotInput) =>
		invoke<boolean>("crypto_verify_password_slot", {
			password: input.password,
			saltB64: input.saltB64,
			slotIdB64: input.slotIdB64,
			verifierB64: input.verifierB64,
			magicVersion: bytes(input.magicVersion),
		}),

	// Wired because the core calls cost nothing, but `securityKeys` is off for desktop in
	// flags.ts: the webview has no usable WebAuthn, so nothing can produce an hmac-secret
	// until there is a native CTAP path.
	wrapVekWebauthn: (input: WrapWebauthnSlotInput) =>
		invoke<PasswordSlotBlob>("crypto_wrap_vek_webauthn", {
			hmacSecretB64: input.hmacSecretB64,
			slotIdB64: input.slotIdB64,
			magicVersion: bytes(input.magicVersion),
		}),
	unwrapVekWebauthn: async (input: UnwrapWebauthnSlotInput) => {
		const ok = await invoke<boolean>("crypto_unwrap_vek_webauthn", {
			hmacSecretB64: input.hmacSecretB64,
			slotIdB64: input.slotIdB64,
			verifierB64: input.verifierB64,
			wrapIvB64: input.wrapIvB64,
			wrappedVekB64: input.wrappedVekB64,
			magicVersion: bytes(input.magicVersion),
		});
		if (ok) markUnlocked();
		return ok;
	},
	verifyWebauthnSlot: (input: VerifyWebauthnSlotInput) =>
		invoke<boolean>("crypto_verify_webauthn_slot", {
			hmacSecretB64: input.hmacSecretB64,
			slotIdB64: input.slotIdB64,
			verifierB64: input.verifierB64,
			magicVersion: bytes(input.magicVersion),
		}),

	encryptEntry: (plaintextJson) =>
		invoke<EncryptedPayload>("crypto_encrypt_entry", { plaintextJson }),
	decryptEntry: (payload) => invoke<string>("crypto_decrypt_entry", { payload }),
	// One IPC round trip for the whole vault, not one per entry.
	decryptEntries: (payloads) => invoke<string[]>("crypto_decrypt_entries", { payloads }),
	encryptWithVek: (plaintext) => invoke<VekEncrypted>("crypto_encrypt_with_vek", { plaintext }),
	decryptWithVek: (iv, ciphertext) => invoke<string>("crypto_decrypt_with_vek", { iv, ciphertext }),

	// Passkey PROVIDER (mint/assert for other sites) stays unwired, and not just for effort:
	// a Tauri webview has no WebAuthn surface to intercept and no OS registration, so there is
	// nothing here to serve. macOS gets it through an ASCredentialProviderExtension instead,
	// which is Swift calling this same core. See docs/macos-credential-provider.md.
	passkeyMakeCredential: (): Promise<PasskeyRegistration> => notWired("Passkey provider"),
	passkeyGetAssertion: (): Promise<PasskeyAssertion> => notWired("Passkey assertion"),

	// Importing them is a different question, and the answer is yes: desktop is a full sync
	// peer, so a passkey dropped here never reaches the user's phone or browser either.
	passkeyImportPkcs8: (pkcs8B64) =>
		invoke<PasskeyImportResult>("crypto_passkey_import_pkcs8", { pkcs8B64 }),

	// Rejects with a bare KDBX_* code string (Tauri rejects with the command's error, not an
	// Error object); `kdbxErrorMessage` reads either. Import only: `saveKdbx` stays absent
	// because the whole save half of core-rust's kdbx.rs is `cfg(feature = "wasm")`. Note the
	// Settings export row is gated on `shell.exportBytes`, which desktop HAS, so that button
	// is visible here and can only throw "KDBX export isn't available here."
	openKdbx: (input: OpenKdbxInput) =>
		invoke<KdbxRawEntry[]>("crypto_open_kdbx", {
			fileB64: input.fileB64,
			password: input.password,
			keyfileB64: input.keyfileB64,
		}),
};

File: packages/platform-desktop/src-tauri/src/crypto.rs
//! The vault crypto core, exposed to the webview as Tauri commands.
//!
//! Every command is a thin wrapper over `vault_crypto`, which this crate links as an
//! ordinary cargo dependency. The VEK therefore lives in this process and never crosses
//! into the webview, which is the desktop equivalent of the extension's offscreen document
//! and mobile's native plugin. See docs/desktop-port.md.
//!
//! The core's structs already serialize camelCase, so they land on the TS side matching
//! `@core/adapters/crypto` exactly and need no mapping layer.

use base64::{engine::general_purpose::STANDARD as B64, Engine};
use vault_crypto::{
    CryptoError, EncryptedPayload, MasterEncrypted, OutEntry, PasskeyImportResult, PasswordSlotBlob,
};

/// Errors ride to JS as a plain string, which `invoke` rejects with. The core already
/// sanitizes these (no key material in a message), so passing Display through is safe.
type CmdResult<T> = Result<T, String>;

fn map<T>(r: Result<T, CryptoError>) -> CmdResult<T> {
    r.map_err(|e| e.to_string())
}

// ---- VEK lifecycle ----

#[tauri::command]
pub fn crypto_is_locked() -> bool {
    vault_crypto::is_locked()
}

#[tauri::command]
pub fn crypto_lock() {
    vault_crypto::lock()
}

#[tauri::command]
pub fn crypto_generate_vek() -> CmdResult<String> {
    map(vault_crypto::generate_vek())
}

#[tauri::command]
pub fn crypto_unlock_with_vek(vek_b64: String) -> CmdResult<()> {
    map(vault_crypto::unlock_with_vek(vek_b64))
}

#[tauri::command]
pub fn crypto_export_vek() -> CmdResult<String> {
    map(vault_crypto::export_vek())
}

#[tauri::command]
pub fn crypto_rotate_vek() -> CmdResult<String> {
    map(vault_crypto::rotate_vek())
}

// ---- slot material ----

#[tauri::command]
pub fn crypto_generate_salt() -> CmdResult<String> {
    map(vault_crypto::generate_salt())
}

#[tauri::command]
pub fn crypto_generate_slot_id() -> CmdResult<String> {
    map(vault_crypto::generate_slot_id())
}

// ---- password slots ----

#[tauri::command]
pub fn crypto_wrap_vek_password(
    password: String,
    salt_b64: String,
    slot_id_b64: String,
    magic_version: Vec<u8>,
) -> CmdResult<PasswordSlotBlob> {
    map(vault_crypto::wrap_vek_password(
        password,
        salt_b64,
        slot_id_b64,
        magic_version,
    ))
}

#[tauri::command]
pub fn crypto_unwrap_vek_password(
    password: String,
    salt_b64: String,
    slot_id_b64: String,
    verifier_b64: String,
    wrap_iv_b64: String,
    wrapped_vek_b64: String,
    magic_version: Vec<u8>,
) -> CmdResult<bool> {
    map(vault_crypto::unwrap_vek_password(
        password,
        salt_b64,
        slot_id_b64,
        verifier_b64,
        wrap_iv_b64,
        wrapped_vek_b64,
        magic_version,
    ))
}

#[tauri::command]
pub fn crypto_verify_password_slot(
    password: String,
    salt_b64: String,
    slot_id_b64: String,
    verifier_b64: String,
    magic_version: Vec<u8>,
) -> CmdResult<bool> {
    map(vault_crypto::verify_password_slot(
        password,
        salt_b64,
        slot_id_b64,
        verifier_b64,
        magic_version,
    ))
}

// ---- security-key slots ----
// Wired because the core calls are free, though `securityKeys` is off for desktop in
// flags.ts until there is a native CTAP path (the webview has no usable WebAuthn).

#[tauri::command]
pub fn crypto_wrap_vek_webauthn(
    hmac_secret_b64: String,
    slot_id_b64: String,
    magic_version: Vec<u8>,
) -> CmdResult<PasswordSlotBlob> {
    map(vault_crypto::wrap_vek_webauthn(
        hmac_secret_b64,
        slot_id_b64,
        magic_version,
    ))
}

#[tauri::command]
pub fn crypto_unwrap_vek_webauthn(
    hmac_secret_b64: String,
    slot_id_b64: String,
    verifier_b64: String,
    wrap_iv_b64: String,
    wrapped_vek_b64: String,
    magic_version: Vec<u8>,
) -> CmdResult<bool> {
    map(vault_crypto::unwrap_vek_webauthn(
        hmac_secret_b64,
        slot_id_b64,
        verifier_b64,
        wrap_iv_b64,
        wrapped_vek_b64,
        magic_version,
    ))
}

#[tauri::command]
pub fn crypto_verify_webauthn_slot(
    hmac_secret_b64: String,
    slot_id_b64: String,
    verifier_b64: String,
    magic_version: Vec<u8>,
) -> CmdResult<bool> {
    map(vault_crypto::verify_webauthn_slot(
        hmac_secret_b64,
        slot_id_b64,
        verifier_b64,
        magic_version,
    ))
}

// ---- entries ----

#[tauri::command]
pub fn crypto_encrypt_entry(plaintext_json: String) -> CmdResult<EncryptedPayload> {
    map(vault_crypto::encrypt_entry(plaintext_json))
}

#[tauri::command]
pub fn crypto_decrypt_entry(payload: EncryptedPayload) -> CmdResult<String> {
    map(vault_crypto::decrypt_entry(
        payload.ciphertext,
        payload.iv,
        payload.wrapped_dek,
        payload.dek_iv,
    ))
}

/// Batch form: one IPC round trip for the whole vault rather than one per entry, which is
/// what dominates open time on large vaults. Order matches `payloads`.
#[tauri::command]
pub fn crypto_decrypt_entries(payloads: Vec<EncryptedPayload>) -> CmdResult<Vec<String>> {
    payloads
        .into_iter()
        .map(|p| {
            map(vault_crypto::decrypt_entry(
                p.ciphertext,
                p.iv,
                p.wrapped_dek,
                p.dek_iv,
            ))
        })
        .collect()
}

#[tauri::command]
pub fn crypto_encrypt_with_vek(plaintext: String) -> CmdResult<MasterEncrypted> {
    map(vault_crypto::encrypt_with_vek(plaintext))
}

#[tauri::command]
pub fn crypto_decrypt_with_vek(iv: String, ciphertext: String) -> CmdResult<String> {
    map(vault_crypto::decrypt_with_vek(iv, ciphertext))
}

// ---- Foreign-format import ----
//
// Both are pure conversions with no VEK involved: the caller encrypts what comes back
// through the ordinary entry path. They are commands rather than webview code only because
// the parsers live in this crate. See docs/passkey-import.md and docs/vault-format.md.

/// Convert a foreign P-256 PKCS#8 passkey into the scalar + COSE pair we store. Without
/// this every Bitwarden passkey is dropped on import, which is silent data loss at exactly
/// the moment a user is migrating in.
#[tauri::command]
pub fn crypto_passkey_import_pkcs8(pkcs8_b64: String) -> CmdResult<PasskeyImportResult> {
    map(vault_crypto::passkey_import_pkcs8(pkcs8_b64))
}

/// Open a KDBX4 database. The file rides as base64 rather than a byte array because Tauri's
/// IPC is JSON, where a `Vec<u8>` is a list of decimal numbers (~4x the bytes for a whole
/// database), and the webview already holds it as base64.
#[tauri::command]
pub fn crypto_open_kdbx(
    file_b64: String,
    password: String,
    keyfile_b64: Option<String>,
) -> CmdResult<Vec<OutEntry>> {
    let file = decode_b64(&file_b64, "file")?;
    let keyfile = keyfile_b64
        .map(|k| decode_b64(&k, "key file"))
        .transpose()?;
    map(vault_crypto::open_kdbx4(file, password, keyfile))
}

/// Decode an IPC base64 argument. A failure here is a bug on our side, not a bad database,
/// so it deliberately carries no `KDBX_*` code: the UI must not blame the user's file.
fn decode_b64(value: &str, what: &str) -> CmdResult<Vec<u8>> {
    B64.decode(value)
        .map_err(|_| format!("could not decode the {what} sent over IPC"))
}
